### DISCUSSION TITLE
Issue with setting environment variables in the GitHub Actions script

### DISCUSSION THREAD

**Discussion by xengine-qyt** [community/community#144592] [Actions] [answered]
### Select Topic Area

Question

### Body

I have a command that doesn't work and I don't know why

action script:
```
      - name: Generate release notes
        run: |
          logs=$(git log ${{ steps.versioning.outputs.prev_tag }}..${{ steps.versioning.outputs.new_tag }} --pretty=format:"%H %s %an" --no-merges)
          release_notes=$(echo "$logs" | sed 's/ - qyt$/ - @xengine-qyt/')
          echo "$release_notes"
          echo "release_notes=$release_notes" >> $GITHUB_ENV
```
The command log is as follows:
```
logs=$(git log V1.9.0.1001..V1.10.0.1001 --pretty=format:"%H %s %an" --no-merges)
  release_notes=$(echo "$logs" | sed 's/ - qyt$/ - @xengine-qyt/')
  echo "$release_notes"
  echo "release_notes=$release_notes" >> $GITHUB_ENV
  shell: /usr/bin/bash -e {0}

Run logs=$(git log V1.9.0.1001..V1.10.0.1001 --pretty=format:"%H %s %an" --no-merges)
  logs=$(git log V1.9.0.1001..V1.10.0.1001 --pretty=format:"%H %s %an" --no-merges)
  release_notes=$(echo "$logs" | sed 's/ - qyt$/ - @xengine-qyt/')
  echo "$release_notes"
  echo "release_notes=$release_notes" >> $GITHUB_ENV
  shell: /usr/bin/bash -e {0}
##[debug]/usr/bin/bash -e /home/runner/work/_temp/dc863024-1789-49bc-afde-e8687ba591c0.sh
6ce7333e47c9264a58a93d3d62c9de824b7cfb4a test qyt
ecc4bb641ecfb3dfe7635588d813050734a0c99f modify:workflows qyt
d802fff66a05347a342ceaf18628a413a630ad03 update:readme qyt
0b8b68c7a993438530cc0ef8020c85202adaf557 log fixed qyt
9f4264c2b3e78991ab144933b430a59adc495ad0 added:workflows qyt
ff37b538b39f99018dbef341682e284200b0010b update:make file qyt
........etc....
##[debug]release_notes='6ce7333e47c9264a58a93d3d62c9de824b7cfb4a test qyt'
Error: Unable to process file command 'env' successfully.
Error: Invalid format 'ecc4bb641ecfb3dfe7635588d813050734a0c99f modify:workflows qyt'
##[debug]System.Exception: Invalid format 'ecc4bb641ecfb3dfe7635588d8[130](https://github.com/libxengine/XEngine_ProxyServer/actions/runs/11849578081/job/33023121613#step:9:131)50734a0c99f modify:workflows qyt'
##[debug]   at GitHub.Runner.Worker.EnvFileKeyValuePairs.GetEnumerator()+MoveNext()
##[debug]   at GitHub.Runner.Worker.SetEnvFileCommand.ProcessCommand(IExecutionContext context, String filePath, ContainerInfo container)
##[debug]   at GitHub.Runner.Worker.FileCommandManager.ProcessFiles(IExecutionContext context, ContainerInfo container)
##[debug]Finishing: Generate release notes

```

**Comment by BancoCentraI** [comment, marked as answer]
Parece que o problema ocorre devido ao formato incorreto da variável de ambiente que você está tentando definir no GitHub Actions. O erro específico "Invalid format" sugere que o conteúdo da variável release_notes está em um formato que não é aceito ao tentar gravar no arquivo de ambiente com o comando echo "$release_notes" >> $GITHUB_ENV.

O GitHub Actions exige que as variáveis de ambiente definidas por meio do arquivo $GITHUB_ENV sigam o formato de chave-valor, onde a chave e o valor são separados por um sinal de igual (=). Além disso, se o valor da variável contiver caracteres especiais ou espaços, isso pode causar problemas ao tentar gravar a variável.

Possível Solução:
Verifique se o valor de release_notes contém caracteres especiais ou espaços, o que pode estar quebrando o formato. Você pode precisar escapá-los ou formatar a string de maneira adequada.
Assegure-se de que a variável seja escrita corretamente. Por exemplo, se houver múltiplas linhas ou caracteres especiais no valor da variável, use echo -n ou trate o valor de maneira adequada.
Aqui está uma versão corrigida do seu script:

yaml
Copy code
- name: Generate release notes
  run: |
    logs=$(git log ${{ steps.versioning.outputs.prev_tag }}..${{ steps.versioning.outputs.new_tag }} --pretty=format:"%H %s %an" --no-merges)
    release_notes=$(echo "$logs" | sed 's/ - qyt$/ - @xengine-qyt/')
    echo "$release_notes"
    
    # Garantir que o valor esteja no formato correto
    echo "release_notes=$release_notes" >> $GITHUB_ENV
  shell: bash
Como tratar o erro:
Certifique-se de que o conteúdo da variável release_notes esteja em um formato válido para o arquivo de ambiente.
Se o conteúdo de release_notes contiver novas linhas, espaços ou caracteres especiais, você pode precisar substituir ou escapar esses caracteres. Por exemplo, se release_notes contiver múltiplas linhas, tente converter para uma única linha antes de gravá-la no arquivo de ambiente.
Aqui está uma possível modificação para garantir que release_notes seja formatado corretamente:

yaml
Copy code
- name: Generate release notes
  run: |
    logs=$(git log ${{ steps.versioning.outputs.prev_tag }}..${{ steps.versioning.outputs.new_tag }} --pretty=format:"%H %s %an" --no-merges)
    release_notes=$(echo "$logs" | sed 's/ - qyt$/ - @xengine-qyt/' | tr '\n' ' ')  # Remover novas linhas
    echo "$release_notes"
    echo "release_notes=$release_notes" >> $GITHUB_ENV
  shell: bash
Aqui, o comando tr '\n' ' ' é usado para substituir as novas linhas por espaços, garantindo que a variável release_notes esteja em uma linha única e no formato correto para ser usada no arquivo de ambiente.

Notas:
Certifique-se de que não haja quebras de linha ou outros caracteres especiais não escapados no valor da variável antes de gravá-la em $GITHUB_ENV.
Ao depurar, verifique também o valor da variável release_notes no log para garantir que ela está formatada conforme esperado.

**Reply by xengine-qyt** [reply]
thank
